builtins/web/fetch/fetch-utils.cpp

```diff
@@ -1,11 +1,147 @@
 #include "fetch-utils.h"
 
-#include <string>
 #include <charconv>
+#include <string>
+#include <ranges>
+#include <string_view>
+#include <fmt/format.h>
 
 namespace builtins::web::fetch {
 
-std::optional<std::tuple<size_t, size_t>> extract_range(std::string_view range_query, size_t full_len) {
+std::string MimeType::to_string() {
+  std::string result = essence;
+
+  for (const auto& [key, value] : params) {
+      result += fmt::format(";{}={}", key, value);
+  }
+
+  return result;
+}
+
+std::string_view trim(std::string_view input) {
+    auto trimsz = input.find_first_not_of(" \t");
+    if (trimsz == std::string_view::npos) {
+        return {};
+    }
+
+    input.remove_prefix(trimsz);
+
+    trimsz = input.find_last_not_of(" \t");
+    input.remove_suffix(input.size() - trimsz - 1);
+
+    return input;
+}
+
+std::optional<MimeType> parse_mime_type(std::string_view str) {
+  auto input = trim(str);
+
+  if (input.empty()) {
+    return std::nullopt;
+  }
+
+  std::string essence;
+  std::string params;
+
+  if (auto pos = input.find(';'); pos != std::string::npos) {
+    essence = trim(input.substr(0, pos));
+    params = trim(input.substr(pos + 1));
+  } else {
+    essence = trim(input);
+  }
+
+  if (essence.empty() || essence.find('/') == std::string::npos) {
+    return std::nullopt;
+  }
+
+  MimeType mime;
+  mime.essence = essence;
+
+  if (params.empty()) {
+    return mime;
+  }
+
+  auto as_string = [](std::string_view v) -> std::string {
+    return {v.data(), v.size()};
+  };
+
+  for (const auto view : std::views::split(params, ';')) {
+    auto param = std::string_view(view.begin(), view.end());
+
+    if (auto pos = param.find('='); pos != std::string::npos) {
+      auto key = trim(param.substr(0, pos));
+      auto value = trim(param.substr(pos + 1));
+
+      if (!key.empty()) {
+        mime.params.push_back({ as_string(key), as_string(value) });
+      }
+    }
+  }
+
+  return mime;
+}
+
+// https://fetch.spec.whatwg.org/#concept-body-mime-type
+std::optional<MimeType> extract_mime_type(std::string_view query) {
```

Comment:
Maybe change this to using `mozilla::Result`, given that a `null` value signals an error?